ml-agents/mlagents/trainers/ghost/trainer.py

```diff
@@ -0,0 +1,259 @@
+            except AgentManagerQueue.Empty:
+                pass
+
+        self.next_update_step = self.trainer.next_update_step
+        self.trainer.advance()
+        self._maybe_write_summary(self.get_step)
+
+        for internal_q in self.internal_policy_queues:
+            # Get policies that correspond to the policy queue in question
+            try:
+                policy = cast(TFPolicy, internal_q.get_nowait())
+                with policy.graph.as_default():
+                    weights = policy.tfvars.get_weights()
+                    self.current_policy_snapshot = weights
+                self.learning_policy_queues[internal_q.behavior_id].put(policy)
+            except AgentManagerQueue.Empty:
+                pass
+
+        if self.get_step - self.last_step > self.steps_between_snapshots:
+            self.save_snapshot(self.trainer.policy)
+            self.last_step = self.get_step
+            self.swap_snapshots()
+
+    def end_episode(self):
+        self.trainer.end_episode()
+
+    def save_model(self, name_behavior_id: str) -> None:
+        self.trainer.save_model(name_behavior_id)
+
+    def export_model(self, name_behavior_id: str) -> None:
+        self.trainer.export_model(name_behavior_id)
+
+    def create_policy(self, brain_parameters: BrainParameters) -> TFPolicy:
+        return self.trainer.create_policy(brain_parameters)
+
+    def add_policy(self, name_behavior_id: str, policy: TFPolicy) -> None:
+        # for saving/swapping snapshots
+        with policy.graph.as_default():
+            policy.tfvars = TensorFlowVariables(policy.model.output, policy.sess)
+
+        self.policies[name_behavior_id] = policy
+
+        # First policy encountered
+        if not self.learning_behavior_name:
+            with policy.graph.as_default():
+                weights = policy.tfvars.get_weights()
+                self.current_policy_snapshot = weights
+            for i in range(self.window):
+                self.policy_snapshots[i] = weights
+            self.trainer.add_policy(name_behavior_id, policy)
+            self.learning_behavior_name = name_behavior_id
+
+    def get_policy(self, name_behavior_id: str) -> TFPolicy:
+        return self.policies[name_behavior_id]
+
+    def save_snapshot(self, policy: TFPolicy) -> None:
+        with policy.graph.as_default():
+            weights = policy.tfvars.get_weights()
+            self.policy_snapshots[self.snapshot_counter] = weights
+        self.policy_elos[self.snapshot_counter] = self.current_elo
+        self.snapshot_counter = (self.snapshot_counter + 1) % self.window
+
+    def swap_snapshots(self) -> None:
+        for q in self.policy_queues:
+            name_behavior_id = q.behavior_id
+            # here is the place for a sampling protocol
+            if name_behavior_id == self.learning_behavior_name:
+                continue
+            elif np.random.uniform() < (1 - self.play_against_current_self_ratio):
```

Comment:
Is this seeded at all?  Does it matter if it is or isn't?

Reply:
For reproducibility, seeding would matter but for general training it shouldn't. Numpy gets seeded if a seed is specified at the beginning of a run.

Reply:
numpy gets seeded here https://github.com/Unity-Technologies/ml-agents/blob/48e21adcfed79ead46a95c37f96034abedd3328f/ml-agents/mlagents/trainers/trainer_controller.py#L67